openshift/odo
# Create deployment based on Devfile

**What type of PR is this?**
/kind feature

## What does this PR do / why we need it
Implements Deployment creation when the `push` command is used.

## List of changes
- Added convenience functions to devfile parser package to get devfile components
- Updated container generator to accept resource requirements (eg. memoryLimit)
- Implemented utility functions for converting environment variables and resource requirements from Devfile types to kube api types
- Removed service account property from pod spec for now because it is optional and we don't have a way to determine whether a non-default service account should be used at this point in time
- Created a devfile command context that can be applicable for any devfile-related commands, it currently only contains the component name
- Created a platform adapter interface that currently only defines a Start() function
- Implemented a Kubernetes platform adapter that creates a Deployment with a fat pod representing the devfile components
- Created a common platform package that contains functions common to all platforms
- Created helper functions for creating a platform adapter based on a Devfile
- Updated existing tests
- Updated component name generation to ensure the name satisfies DNS-1123 rules
- New tests
- Rebased on master and updated to work with https://github.com/openshift/odo/pull/2598

**Which issue(s) this PR fixes**:

Fixes #2471

## How to test changes / Special notes to the reviewer**:

**Automated tests**
Run the kclient tests:
In `odo/pkg/kclient` run `go test ./... -v`

Expected results:
```
=== RUN   TestCreateDeployment
=== RUN   TestCreateDeployment/Case:_Valid_deployment_name
=== RUN   TestCreateDeployment/Case:_Invalid_deployment_name
--- PASS: TestCreateDeployment (0.00s)
    --- PASS: TestCreateDeployment/Case:_Valid_deployment_name (0.00s)
    --- PASS: TestCreateDeployment/Case:_Invalid_deployment_name (0.00s)
=== RUN   TestUpdateDeployment
=== RUN   TestUpdateDeployment/Case:_Valid_deployment_name
=== RUN   TestUpdateDeployment/Case:_Invalid_deployment_name
--- PASS: TestUpdateDeployment (0.00s)
    --- PASS: TestUpdateDeployment/Case:_Valid_deployment_name (0.00s)
    --- PASS: TestUpdateDeployment/Case:_Invalid_deployment_name (0.00s)
=== RUN   TestGenerateContainer
=== RUN   TestGenerateContainer/#00
=== RUN   TestGenerateContainer/container1
--- PASS: TestGenerateContainer (0.00s)
    --- PASS: TestGenerateContainer/#00 (0.00s)
    --- PASS: TestGenerateContainer/container1 (0.00s)
=== RUN   TestGeneratePodTemplateSpec
=== RUN   TestGeneratePodTemplateSpec/podSpecTest
--- PASS: TestGeneratePodTemplateSpec (0.00s)
    --- PASS: TestGeneratePodTemplateSpec/podSpecTest (0.00s)
=== RUN   TestGeneratePVCSpec
=== RUN   TestGeneratePVCSpec/1
=== RUN   TestGeneratePVCSpec/2
--- PASS: TestGeneratePVCSpec (0.00s)
    --- PASS: TestGeneratePVCSpec/1 (0.00s)
    --- PASS: TestGeneratePVCSpec/2 (0.00s)
=== RUN   TestWaitAndGetPod
=== RUN   TestWaitAndGetPod/phase:_running
 •  Waiting for component to start  ...
 ✓  Waiting for component to start [200661ns]
=== RUN   TestWaitAndGetPod/phase:_failed
 •  Waiting for component to start  ...
 ✗  Waiting for component to start [88055ns]
=== RUN   TestWaitAndGetPod/phase:_unknown
 •  Waiting for component to start  ...
 ✗  Waiting for component to start [56931ns]
--- PASS: TestWaitAndGetPod (0.00s)
    --- PASS: TestWaitAndGetPod/phase:_running (0.00s)
    --- PASS: TestWaitAndGetPod/phase:_failed (0.00s)
    --- PASS: TestWaitAndGetPod/phase:_unknown (0.00s)
=== RUN   TestCreatePVC
=== RUN   TestCreatePVC/Case:_Valid_pvc_name
=== RUN   TestCreatePVC/Case:_Invalid_pvc_name
--- PASS: TestCreatePVC (0.00s)
    --- PASS: TestCreatePVC/Case:_Valid_pvc_name (0.00s)
    --- PASS: TestCreatePVC/Case:_Invalid_pvc_name (0.00s)
=== RUN   TestAddPVCToPodTemplateSpec
=== RUN   TestAddPVCToPodTemplateSpec/podSpecTest
--- PASS: TestAddPVCToPodTemplateSpec (0.00s)
    --- PASS: TestAddPVCToPodTemplateSpec/podSpecTest (0.00s)
=== RUN   TestAddVolumeMountToPodTemplateSpec
=== RUN   TestAddVolumeMountToPodTemplateSpec/podSpecTest
=== RUN   TestAddVolumeMountToPodTemplateSpec/podSpecTest#01
--- PASS: TestAddVolumeMountToPodTemplateSpec (0.00s)
    --- PASS: TestAddVolumeMountToPodTemplateSpec/podSpecTest (0.00s)
    --- PASS: TestAddVolumeMountToPodTemplateSpec/podSpecTest#01 (0.00s)
=== RUN   TestGetPVCsFromSelector
=== RUN   TestGetPVCsFromSelector/Case:_Valid_pvc_name
=== RUN   TestGetPVCsFromSelector/Case:_Wrong_Label_Selector
--- PASS: TestGetPVCsFromSelector (0.00s)
    --- PASS: TestGetPVCsFromSelector/Case:_Valid_pvc_name (0.00s)
    --- PASS: TestGetPVCsFromSelector/Case:_Wrong_Label_Selector (0.00s)
PASS
ok  	github.com/openshift/odo/pkg/kclient	0.446s
```

Run the new adapter tests:
In `odo/pkg/devfile/adapters` run `go test ./... -v`

Expected results:
```
=== RUN   TestNewPlatformAdapter
=== RUN   TestNewPlatformAdapter/get_platform_adapter
--- PASS: TestNewPlatformAdapter (0.00s)
    --- PASS: TestNewPlatformAdapter/get_platform_adapter (0.00s)
PASS
ok  	github.com/openshift/odo/pkg/devfile/adapters	(cached)
?   	github.com/openshift/odo/pkg/devfile/adapters/common	[no test files]
?   	github.com/openshift/odo/pkg/devfile/adapters/kubernetes	[no test files]
=== RUN   TestStart
=== RUN   TestStart/Case:_Invalid_devfile
=== RUN   TestStart/Case:_Valid_devfile
Successfully created component test
--- PASS: TestStart (0.00s)
    --- PASS: TestStart/Case:_Invalid_devfile (0.00s)
    --- PASS: TestStart/Case:_Valid_devfile (0.00s)
PASS
ok  	github.com/openshift/odo/pkg/devfile/adapters/kubernetes/component	(cached)
```

**Manual test instructions**
1. Use experimental mode by either:
    - setting an env variable "ODO_EXPERIMENTAL=true" OR
    - setting odo preference using `odo preference set experimental true`

1. Clone https://github.com/rajivnathan/openLibertyDevfile

1. Run `odo push`

1. Verify the following are created
    - a deployment
    - a pod with a container for each devfile component of `DockerImage` type and has an `alias`
